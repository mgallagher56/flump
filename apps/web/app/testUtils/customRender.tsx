// create a custom render function with ChakraProvider
import { type ReactNode } from 'react';

import { type RenderOptions, type RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { Provider } from '~/components/ui/provider';

const user = userEvent.setup();

const customRender = (ui: ReactNode, options?: RenderOptions): RenderResult & { user: UserEvent } => {
  const wrappedUi = <Provider>{ui}</Provider>;

  return {
    ...rtlRender(wrappedUi, options),
    user
  };
};

export default customRender;
