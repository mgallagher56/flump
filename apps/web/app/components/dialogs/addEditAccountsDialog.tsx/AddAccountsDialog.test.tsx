import type { ReactNode } from 'react';

import { vi } from 'vitest';
import customRender from '~/testUtils/customRender';

import mockUser from '__mocks__/user';

import AddEditAccountsDialogBtn from './AddEditAccountsDialog';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn(),
  mockUseRevalidator: vi.fn(() => ({ revalidate: vi.fn() })),
  mockFrom: vi.fn(() => ({
    insert: () => ({
      eq: () => ({
        eq: () => ({})
      }),
      select: vi.fn(() => ({
        data: [{ id: '123456' }]
      }))
    })
  }))
}));

vi.mock('react-router', async () => {
  const actual: Record<string, unknown> = await vi.importActual('react-router');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData,
    useRevalidator: mocks.mockUseRevalidator,
    Form: ({ children }: { children: ReactNode }) => <form>{children}</form>,
    useSubmit: () => ({ onSubmit: vi.fn() })
  };
});

vi.mock('app/utils/supabase', () => ({
  default: {
    from: mocks.mockFrom
  }
}));

describe('<AddAccountDialogBtn', () => {
  test('should render add account dialog when trigger button is clicked', async () => {
    mocks.mockUseLoaderData.mockReturnValue({ user: mockUser });
    const { baseElement, getByText, getAllByText, user } = customRender(
      <AddEditAccountsDialogBtn accountId="123456" />
    );
    const triggerBtn = getByText('addAccount');
    expect(triggerBtn).toBeDefined();
    await user.click(triggerBtn);

    const nameInput = getByText('Name');
    expect(nameInput).toBeDefined();
    const accountTypeInput = getByText('Account Type');
    expect(accountTypeInput).toBeDefined();

    await user.type(nameInput, 'Starling');

    await user.type(accountTypeInput, 'Saving');

    const addAccountBtn = getAllByText('addAccount')[2];
    expect(addAccountBtn).toBeDefined();
    await user.click(addAccountBtn);
    expect(mocks.mockFrom).toBeCalled();
    const htmlString = baseElement.outerHTML.toString();
    const baseElementConstant = htmlString.replaceAll(/style="[^"]*"/g, '');
    expect(baseElementConstant).toMatchSnapshot();
  });
});
