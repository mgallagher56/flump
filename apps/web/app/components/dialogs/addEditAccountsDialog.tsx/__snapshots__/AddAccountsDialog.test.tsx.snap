// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AddAccountDialogBtn > should render add account dialog when trigger button is clicked 1`] = `"<body><div><script nonce="">((e,r,s,u,d,m,l,h)=>{let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=>{let k=y==="class",S=k&&m?d.map(f=>m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)</script><button data-scope="dialog" data-part="trigger" dir="ltr" id="dialog::r0::trigger" aria-haspopup="dialog" type="button" aria-expanded="false" data-state="closed" class="chakra-dialog__trigger css-0"><button type="button" class="chakra-button css-teolye">addAccount</button></button></div></body>"`;
